### ISSUE TITLE
Gen 2 architecture: what porting Needle to a second runtime cost, and what would make the third one cheap

### ISSUE THREAD

**Issue by noelsaw1** [HiQS-Labs/Needle-fork#7]
Architectural follow-up from Phase 2 ([#1](https://github.com/HiQS-Labs/Needle-fork/issues/1)) and the MLX port ([#5](https://github.com/HiQS-Labs/Needle-fork/issues/5)). Not a bug report and not a request to change Gen 1 — a record of what we hit when we tried to move Needle to a second execution substrate, written while the evidence is fresh so a Gen 2 design can price it in.

Everything below is measured on this fork against `checkpoints/needle2.pkl` (27 layers, 8 heads, `d_model` 512, 4 MHC lanes). Where a claim is an opinion it is marked as one.

## Why this is worth writing down now

Phase 2 needed Needle to train somewhere other than JAX-on-CPU, because JAX on Apple Silicon *is* CPU (`jax-metal` does not work past jax 0.4.38, `doc/finetuning.md:63`; a current install reports `Available backends are ['cpu']`) and CPU training measured **~127 h/epoch**.

So we ported the architecture to MLX. It worked — parity is `4.172e-07` max |Δlogits| on a tiny fixture and `2.669e-05` relative on the real checkpoint, argmax 100%, top-3 99.98% — but the cost of getting there is the interesting part, and most of it was structural rather than incidental.

## 1. The model definition *is* the runtime; there is no portable representation

`architecture.py` is a Flax module. It is simultaneously the specification of what Needle *is* and the implementation for one backend. There is no intermediate representation between them, and `.cact` is not one — `write_export` packs weights, config and tokenizer into an archive. It is a **weights format, not a program**.

The consequence is that a second substrate is not a recompile, it is a **reimplementation**: ~460 hand-written lines reproducing ZCRMSNorm, RoPE, Engram, GQA attention, the Hadamard MLP, Block, and the MHC/Sinkhorn lane mixing in another array library. And because it is a reimplementation rather than a translation, correctness is not inherited — it has to be *proved*, which means a parity harness and a falsification harness are mandatory companions, not nice-to-haves. Ours caught two real bugs (§4 below) that no amount of code review had.

Worth being precise about the ladder, because "port it to X" hides a 10x cost difference:

| | What it is | Cost |
|---|---|---|
| `jax-metal` | genuine recompile — same source, XLA emits Metal | backend flag + a version pin |
| MLX | rewrite in another array library | ~460 lines + parity + falsification harness |
| CoreML / ANE | conversion to a different graph format | unknown, plus op-set risk (§5) |

**Opinion:** the single highest-leverage Gen 2 change is to make the block algebra expressible independently of the array library — even informally, as a spec + conformance suite that any backend must pass — so the third backend costs materially less than the second.

## 2. Memory-bounded attention is a per-runtime reimplementation, and its absence is catastrophic rather than degrading

`TransformerConfig.flash` defaults `True` (`architecture.py:71`) and the JAX path uses it. Our MLX port does not implement it: it materialises the full `[B, H, S, S]` score matrix and softmaxes (`san_mlx.py:214`).

That single gap is the difference between a working configuration and an unusable one. At batch 16 / seq 2048 on 27 layers, retained scores alone are:

```
batch  2 :  0.25 GB/layer x 27 =   6.8 GB
batch  8 :  1.00 GB/layer x 27 =  27.0 GB
batch 16 :  2.00 GB/layer x 27 =  54.0 GB
```

**This took a 24 GB machine down hard** — not an OOM, a reboot. MLX allocates from unified memory with no default ceiling, so the GPU driver starved the kernel. `mx.set_memory_limit` did **not** prevent it: with the cap at 12 GB a step still peaked at 13.4 GB and allocated straight through without raising. The only thing that reliably protects the host is refusing the config before allocating, from arithmetic.

Two Gen 2 asks fall out:

- Treat memory-bounded attention as **part of the architecture contract**, not a per-backend optimisation flag. A backend that cannot do it should be unable to claim conformance, because the failure mode is not "slower", it is "takes the machine down".
- Ship the activation-memory estimate as a **library function** derived from the config. Every runtime needs it, it is a dozen lines, and each port currently rediscovers it the expensive way.

## 3. Quantisation numerics are entangled with the training path, and the deploy-time guard is opt-in

`needle finetune` defaults to `--qat-bits auto`, which on this checkpoint resolves to `CQ mixed[embedding=4,mhc=4,default=2] STE + A8`. So the *default* training objective is quantisation-aware, and `quantize.py`'s codebooks are shared bit-for-bit with the `.cact` export path so that training and deployment numerics agree.

That is a good property, and it is exactly what makes a second runtime expensive: an MLX (or any) training path must reimplement the Lloyd-Max codebook, the Hadamard rotation and the nearest-codeword search, or it silently optimises a different objective than the one it deploys.

The sharper problem is the guard. `build_main` reads `qat_bits` / `qat_bits_map` **out of the adapter pickle** and raises on a conflicting `--bits`. An adapter that simply omits that metadata gets no such check — so a full-precision-trained adapter can be built into quantised numerics it was never trained for, and **nothing raises**. The safety property depends on the writer volunteering metadata.

**Ask:** make the adapter record its training numerics as a required field, and have `build` refuse an adapter that does not declare them, rather than treating silence as "no constraint".

## 4. Dtype policy at block boundaries is implicit, and that is where the port bugs were

Both bugs the parity gate caught were fp16-only, invisible at fp32, and both were the same class: *where* a cast happens relative to an arithmetic op.

- `ZCRMSNorm`: `(1 + scale)` evaluates in the **parameter's** dtype, not float32. Promote first and the result differs.
- `attn_gate`: `sigmoid` must be applied then cast, not cast then applied.

These are not exotic. They are load-bearing semantics carried implicitly by JAX's promotion rules, invisible in the source, and only discoverable by differential testing against a reference. The real checkpoint stores **float16** params (casting to fp32 drops the residual to `4.1e-08` relative — machine-exact), so fp32-only testing would have shipped both.

**Ask:** make dtype policy explicit at block boundaries in the Gen 2 spec — state where accumulation is fp32 and where the parameter dtype governs — so it is a documented contract rather than an emergent property of one library's promotion table.

## 5. Op-set portability is not currently a design constraint, and Phases 3-4 need it to be

Needle's distinctive blocks are also its least portable ones. For a restricted accelerator op-set (ANE via CoreML being the concrete Phase 3-4 target), the ones we would expect trouble from:

- **Engram** — uint32 n-gram hashing plus a **data-dependent table gather**. Dynamic indexing is the classic thing restricted accelerators reject.
- **Hadamard MLP** — Walsh matrices.
- **Sinkhorn** — an iterative normalisation in the MHC residual mixing path.

Nobody has attempted the conversion, so this is a **flagged unknown, not a known blocker**. But it is the thing to find out early, because it is architectural: if the ANE cannot express Engram's gather, that is a Gen 2 design question, not a porting detail. One point in our favour is that the checkpoint is already fp16, which suits the ANE (Phase 1 constraint #18: fp32 program I/O is rejected on M1-generation ANE, accepted on M4).

Note for anyone reading this as ANE progress: **MLX cannot reach the ANE.** `mlx.core.DeviceType` is exactly `['cpu', 'gpu']`. The MLX work contributes nothing toward Phase 3.

## 6. Fine-tuning freezes the confidence head

`needle/__init__.py` warns that *"finetuning does not update the confidence head, so scores are uncalibrated for tuned weights"* and then hard-sets `response["confidence"] = None` for any tuned artifact.

For our use case this was nearly fatal: #1 §5 was specified around confidence separation and #1 §6 around a `min_confidence` hook floor, and both are unimplementable on a LoRA fine-tune. We worked around it by measuring **signal separation** from the ranking signal (`run.py:_step_stats`) instead — which is a fine answer, but it is a workaround for an architectural gap.

**Ask:** Gen 2 should either make the confidence head trainable under fine-tuning, or ship a documented post-hoc recalibration path. A model whose calibration signal is unavailable precisely on the artifacts people actually ship is a gap worth closing at the architecture level.

## 7. Oracle-shaped workloads pay for a constant prompt prefix on every row

Specific to how we use Needle, but it is an architecture question rather than a data one.

Every training row embeds all 44 tool schemas inline: **1,383 tokens of schema** against rows that render at 1,524-1,944 tokens total. So roughly **75-80% of every row is byte-identical to every other row**, and we pay full attention cost over it, every row, every step. It is also why `--max-len` cannot rescue CPU training: the length distribution is near-uniform, coverage at both 512 and 1024 is 0.0%, and there is no short-row bucket to exploit.

**Opinion / suggestion:** a Gen 2 that supports a *tool vocabulary* — schemas resolved to learned embeddings or a cacheable prefix, rather than re-tokenised per example — would cut the dominant cost of exactly the constrained tool-selection workload Needle is built for. For a 45M edge model whose whole premise is a fixed tool surface, re-encoding that fixed surface on every forward pass is the most obviously redundant work in the pipeline.

## Explicit non-goals

- Not proposing MLX become a dependency of `main`. It stays behind an explicit go decision, promoted by a normal PR, never inherited from the spike branch.
- Not proposing Gen 1 changes. This is input to a Gen 2 design.
- Not asking anyone to build the CoreML path on the strength of §5. That is a question to answer, not a plan.

## Receipts

- Parity, falsification and memory numbers: `TESTS-RESULTS/2026-09-07-mlx-spike/`
- Port and harnesses: `spike/mlx/` on `spike/mlx-finetune`
- Phase 2 plan and the revised §4: [#1](https://github.com/HiQS-Labs/Needle-fork/issues/1)
- Execution lane and gate results: [#5](https://github.com/HiQS-Labs/Needle-fork/issues/5)
